### PR TITLE
Accumulate loss function reduce=True into accreal instead of real

### LINKED ISSUE #8710: Incorrect calculation for torch.nn.MSELoss
## Issue description
torch.nn.MSELoss seems to get inaccurate calculation when size_average=True and reduce=True

## Code example

For tensors of 0.1's and 0's, the desired output should be 0.01. (0.1 - 0)^2 = 0.01.
I can get correct result by code below:
```
In [10]: preds = torch.ones(5, 68, 64, 64) / 10
In [11]: res = torch.pow((preds - labels), 2)
In [12]: torch.mean(res)
Out[12]: 
tensor(1.00000e-02 *
       1.0000)
```

However, if I do something like below, I get inaccurate results:
```
In [16]: criterion = torch.nn.MSELoss()
In [17]: criterion(labels, preds)
Out[17]: 
tensor(1.00000e-03 *
       9.8371)
```

But when I turned off size_average and reduce and calculate mean manually, I get desired output:
```
In [21]: criterion = torch.nn.MSELoss(False, False)
In [22]: torch.sum(criterion(labels, preds)) / (5*68*64*64)
Out[22]: 
tensor(1.00000e-02 *
       1.0000)
```

I am not quite sure if it is intended to be like that or I am not using it correctly.

## System Info
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.3 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 2.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti

Nvidia driver version: 387.34
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.2
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7.1.4
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip] numpy (1.14.5)
[pip] torch (0.4.0)
[pip] torchvision (0.2.1)
[conda] Could not collect

### PR DESCRIPTION
**Pull request by zou3519**
THNN was accumulating the result of reduction loss functions
into real instead of accreal. This was causing precision issues with
MSELoss.

~~I added precision tests for some of the losses (KLDivLoss, BCELoss,
L1Loss, SmoothL1Loss, MSELoss). The other ones are a bit harder
to test.~~

This patch only fixes MSELoss. Some of the other losses exhibit bad precision as well (because they accumulate into real instead of accreal) and require more investigation. I will open an issue for those.

Fixes #8710

### CODE REVIEW

**Review by soumith — APPROVED** [review]

**Comment by SsnL** [comment]
Nice! The new tests are written with good `*tol` so they fail before this patch, right?

**Comment by zou3519** [comment]
Don't merge this yet, I think some of the tests are going to fail. @SsnL yes I checked that all of precision tests that I added fail before this patch.

**Comment by li-roy** [comment]
Does it make sense to include these tests if the only reason is so that they fail before this patch and pass after it? I feel that they might not be that helpful in the future.

**Comment by SsnL** [comment]
@li-roy They can prevent regression. Haven't we always been adding tests along with fixes?

**Comment by ezyang** [comment]
@zou3519 Don't forget about this patch!

**Comment by zou3519** [comment]
I'll rebase and submit this with only the MSELoss changes. The other losses have similar precision issues but those require more looking into and aren't immediately solved by accumulating into accreal instead of real.

**Comment by zou3519** [comment]
Closing in favor of #9287
